odl/space/base_tensors.py

```diff
@@ -0,0 +1,741 @@
+            Numpy data type specifier.
+        """
+        raise NotImplementedError('abstract method')
+
+    @staticmethod
+    def available_dtypes():
+        """Return the list of data types available in this implementation."""
+        raise NotImplementedError('abstract method')
+
+    @property
+    def element_type(self):
+        """Type of elements in this set: `Tensor`."""
+        return Tensor
+
+
+class Tensor(LinearSpaceElement):
+
+    """Abstract class for representation of `TensorSpace` elements."""
+
+    def __init__(self, space):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : TensorSpace
+            The space to which this tensor belongs.
+        """
+        assert isinstance(space, TensorSpace)
+        self.__space = space
+
+    @property
+    def impl(self):
+        """Name of the implementation back-end of this tensor."""
+        return self.space.impl
+
+    copy = LinearSpaceElement.copy
+
+    def asarray(self, out=None):
+        """Extract the data of this tensor as a Numpy array.
+
+        Parameters
+        ----------
+        out : `numpy.ndarray`
+            Array to write the result to.
+
+        Returns
+        -------
+        asarray : `numpy.ndarray`
+            Numpy array of the same data type and shape as the space.
+            If ``out`` was given, the returned object is a reference
+            to it.
+        """
+        raise NotImplementedError('abstract method')
+
+    def __getitem__(self, indices):
+        """Return ``self[indices]``.
+
+        Parameters
+        ----------
+        indices : index expression
+            Integer, slice or sequence of these, defining the positions
+            of the data array which should be accessed.
+
+        Returns
+        -------
+        values : `TensorSpace.dtype` or `Tensor`
+            The value(s) at the given indices. Note that depending on
+            the implementation, the returned object may be a (writable)
+            view into the original array.
+        """
+        raise NotImplementedError('abstract method')
+
+    def __setitem__(self, indices, values):
+        """Implement ``self[indices] = values``.
+
+        Parameters
+        ----------
+        indices : index expression
+            Integer, slice or sequence of these, defining the positions
+            of the data array which should be written to.
+        values : scalar, `array-like` or `Tensor`
+            The value(s) that are to be assigned.
+
+            If ``index`` is an integer, ``value`` must be a scalar.
+
+            If ``index`` is a slice or a sequence of slices, ``value``
+            must be broadcastable to the shape of the slice.
+        """
+        raise NotImplementedError('abstract method')
+
+    __eq__ = LinearSpaceElement.__eq__
+
+    def __ne__(self, other):
+        """Return ``self != other``."""
+        return not self == other
+
+    def __hash__(self):
```

Comment:
I didn't think much about it here, thought I remembered another place where hashing of a numpy array was needed. But actually, this shouldn't be hashable, so I'll remove it.